src/debug.cc

```diff
@@ -71,19 +98,359 @@ class DebugFile
+    bool forceAlert; ///< debugs() forceAlert flag
+};
+
+// Avoid SBuf for CompiledDebugMessageBody:
+// * SBuf's own debugging may create a lot of reentrant debugging noise.
+// * Debug::Context::buf is an std::string-based STL ostream. Converting its
+//   buf() result to a different kind of string may increase complexity/cost.
+// TODO: Consider switching to a simple fixed-size buffer and a matching stream!
+/// The processed "content" (i.e. the last parameter) part of a debugs() call.
+using CompiledDebugMessageBody = std::string;
+
+/// a fully processed debugs(), ready to be logged
+class CompiledDebugMessage
+{
+public:
+    using Header = DebugMessageHeader;
+    using Body = CompiledDebugMessageBody;
+
+    CompiledDebugMessage(const Header &aHeader, const Body &aBody);
+
+    Header header; ///< debugs() meta-information; reflected in log line prefix
+    Body body; ///< the log line after the prefix (without the newline)
+};
+
+// We avoid PoolingAllocator for CompiledDebugMessages to minimize reentrant
+// debugging noise. This noise reduction has negligible performance overhead
+// because it only applied to early messages, and there are few of them.
+/// debugs() messages captured in LogMessage() call order
+using CompiledDebugMessages = std::deque<CompiledDebugMessage>;
+
+/// a receiver of debugs() messages (e.g., stderr or cache.log)
+class DebugChannel
+{
+public:
+    using EarlyMessages = std::unique_ptr<CompiledDebugMessages>;
+
+    explicit DebugChannel(const char * const aName): name(aName), earlyMessages(new CompiledDebugMessages()) {}
+    virtual ~DebugChannel() = default;
+
+    // no copying or moving or any kind (for simplicity sake and to prevent accidental copies)
+    DebugChannel(DebugChannel &&) = delete;
+
+    /// whether we are still expecting (and buffering) early messages
+    bool collectingEarlyMessages() const { return bool(earlyMessages); }
+
+    /// end early message buffering, logging any saved messages
+    void stopEarlyMessageCollection();
+
+    /// end early message buffering, without logging any saved messages
+    /// \returns (a possibly empty container with) saved messages or nil
+    EarlyMessages releaseEarlyMessages() { return EarlyMessages(earlyMessages.release()); }
+
+    /// Write the message to the channel if the channel accepts (such) messages.
+    /// This writing may be delayed until the channel configuration is settled.
+    void log(const DebugMessageHeader &, const CompiledDebugMessageBody &);
+
+protected:
+    /// output iterator for writing CompiledDebugMessages to a given channel
+    class Logger
+    {
+    public:
+        using difference_type = void;
+        using value_type = void;
+        using pointer = void;
+        using reference = void;
+        using iterator_category = std::output_iterator_tag;
+
+        explicit Logger(DebugChannel &ch): channel(ch) {}
+
+        Logger &operator=(const CompiledDebugMessage &message) {
```

Comment:
```suggestion
        Logger &operator=(const CompiledDebugMessage &message)
        {
```

Reply:
Fixed in 614c518. Similar problems fixed in c2c643e. We need to find a way to automate this.

Reply:
IIRC the current astyle should be able to do it. I do not think we have enabled and tested that ability yet though.